lib/tests/interactive_scripts.py

```diff
@@ -0,0 +1,426 @@
+        tree.session_state = self.session_state
+        return tree
+
+    def script_stopped(self) -> bool:
+        for e in self.events:
+            if e in (
+                ScriptRunnerEvent.SCRIPT_STOPPED_FOR_RERUN,
+                ScriptRunnerEvent.SCRIPT_STOPPED_WITH_COMPILE_ERROR,
+                ScriptRunnerEvent.SCRIPT_STOPPED_WITH_SUCCESS,
+            ):
+                return True
+        return False
+
+
+def script_from_string(path: pathlib.Path, script: str) -> LocalScriptRunner:
+    aligned_script = textwrap.dedent(script)
+    path.write_text(aligned_script)
+    return LocalScriptRunner(str(path))
+
+
+def script_from_filename(script_name: str) -> LocalScriptRunner:
+    script_path = os.path.join(
+        os.path.dirname(__file__), "streamlit", "test_data", script_name
+    )
+    return LocalScriptRunner(script_path)
+
+
+def require_widgets_deltas(runner: LocalScriptRunner, timeout: float = 3) -> None:
+    """Wait for the given ScriptRunner to emit a completion event. If the timeout
+    is reached, the runner will be shutdown and an error will be thrown.
+    """
+
+    t0 = time.time()
+    while time.time() - t0 < timeout:
+        time.sleep(0.1)
+        if runner.script_stopped():
+            return
+
+    # If we get here, the runner hasn't yet completed before our
+    # timeout. Create an error string for debugging.
+    err_string = f"require_widgets_deltas() timed out after {timeout}s)"
+
+    # Shutdown the runner before throwing an error, so that the script
+    # doesn't hang forever.
+    runner.request_stop()
+    runner.join()
+
+    raise RuntimeError(err_string)
+
+
+@dataclass(init=False)
+class Element:
+    type: str
+    proto: ElementProto = field(repr=False)
+    root: Root = field(repr=False)
+
+    def __init__(self, proto: ElementProto, root: Root):
+        self.proto = proto
+        self.root = root
+        ty = proto.WhichOneof("type")
+        assert ty is not None
+        self.type = ty
+
+    def __iter__(self):
+        yield self
+
+    @property
+    def value(self):
+        """FIXME: this does not reflect state changes from interactions
+        after a script run. We can work around this for now by writing to
+        `st.text` and checking that instead.
+        """
+        p = getattr(self.proto, self.type)
+        return p.value
+
+    @property
+    def key(self) -> str | None:
+        return None
+
+    def widget_state(self) -> WidgetState | None:
+        return None
+
+    def run(self) -> Root:
+        return self.root.run()
+
+
+@dataclass(init=False)
+class Text(Element):
+    proto: TextProto
+    root: Root = field(repr=False)
+
+    def __init__(self, proto: TextProto, root: Root):
+        self.proto = proto
+        self.root = root
+
+    @property
+    def value(self) -> str:
+        return self.proto.body
+
+    @property
+    def type(self) -> str:
+        return "text"
+
+
+@dataclass(init=False)
+class Radio(Element):
```

Comment:
Unclear if we should use properties for the protobuf-delegating accessors, or if we should extract them into fields in the constructor. Some perf differences but probably not worth worrying about, so mostly a question of clarity